File: lib/models/Recipe.dart
// lib/models/recipe.dart (업데이트된 코드)

class Recipe {
  final String id;
  final String title;
  final String foodCategory;
  final String cookingCategory;
  final Images images;
  final String tags;
  final String tips;
  final List<Ingredient> ingredients;
  final List<Method> methods;
  final Nutrition nutrition;

  Recipe({
    required this.id,
    required this.title,
    required this.foodCategory,
    required this.cookingCategory,
    required this.images,
    required this.tags,
    required this.tips,
    required this.ingredients,
    required this.methods,
    required this.nutrition,
  });

  factory Recipe.fromJson(Map<String, dynamic> json) {
    return Recipe(
      id: json['_id'],
      title: json['title'],
      foodCategory: json['food_category'],
      cookingCategory: json['cooking_category'],
      images: Images.fromJson(json['images']),
      tags: json['tags'],
      tips: json['tips'],
      ingredients: (json['ingredients'] as List)
          .map((i) => Ingredient.fromJson(i))
          .toList(),
      methods: (json['methods'] as List)
          .map((m) => Method.fromJson(m))
          .toList(),
      nutrition: Nutrition.fromJson(json['nutrition']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "_id": id,
      "title": title,
      "food_category": foodCategory,
      "cooking_category": cookingCategory,
      "images": images.toJson(),
      "tags": tags,
      "tips": tips,
      "ingredients": ingredients.map((i) => i.toJson()).toList(),
      "methods": methods.map((m) => m.toJson()).toList(),
      "nutrition": nutrition.toJson(),
    };
  }

}

class Images {
  final String originalUrl;
  final String localPath;

  Images({required this.originalUrl, required this.localPath});

  factory Images.fromJson(Map<String, dynamic> json) {
    return Images(
      originalUrl: json['original_url'],
      localPath: json['local_path'],
    );
  }
  Map<String, dynamic> toJson() {
    return {
      "original_url": originalUrl,
      "local_path": localPath,
    };
  }
}

class Ingredient {
  final String name;
  final String quantity;

  Ingredient({required this.name, required this.quantity});

  factory Ingredient.fromJson(Map<String, dynamic> json) {
    return Ingredient(
      name: json['name'],
      quantity: json['quantity'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "name": name,
      "quantity": quantity,
    };
  }
}

class Method {
  final String describe;
  final MethodImage image;

  Method({required this.describe, required this.image});

  factory Method.fromJson(Map<String, dynamic> json) {
    return Method(
      describe: json['describe'],
      image: MethodImage.fromJson(json['image']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "describe": describe,
      "image": image.toJson(),
    };
  }
}

class MethodImage {
  final String originalUrl;
  final String localPath;

  MethodImage({required this.originalUrl, required this.localPath});

  factory MethodImage.fromJson(Map<String, dynamic> json) {
    return MethodImage(
      originalUrl: json['original_url'],
      localPath: json['local_path'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "original_url": originalUrl,
      "local_path": localPath,
    };
  }
}

class Nutrition {
  final String calories;
  final String protein;
  final String carbohydrates;
  final String fat;
  final String sodium;

  Nutrition({
    required this.calories,
    required this.protein,
    required this.carbohydrates,
    required this.fat,
    required this.sodium,
  });

  factory Nutrition.fromJson(Map<String, dynamic> json) {
    return Nutrition(
      calories: json['calories'],
      protein: json['protein'],
      carbohydrates: json['carbohydrates'],
      fat: json['fat'],
      sodium: json['sodium'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "calories": calories,
      "protein": protein,
      "carbohydrates": carbohydrates,
      "fat": fat,
      "sodium": sodium,
    };
  }
}

File: ting-rn/src/types/recipe.ts
export interface RecipeImages {
  originalUrl: string;
  localPath: string;
}

export interface Ingredient {
  name: string;
  quantity: string;
}

export interface MethodImage {
  originalUrl: string;
  localPath: string;
}

export interface RecipeMethod {
  describe: string;
  image: MethodImage;
}

export interface Nutrition {
  calories: string;
  protein: string;
  carbohydrates: string;
  fat: string;
  sodium: string;
}

export interface Recipe {
  id: string;
  title: string;
  foodCategory: string;
  cookingCategory: string;
  images: RecipeImages;
  tags: string;
  tips: string;
  ingredients: Ingredient[];
  methods: RecipeMethod[];
  nutrition: Nutrition;
}

export function recipeToJson(recipe: Recipe): Record<string, unknown> {
  return {
    _id: recipe.id,
    title: recipe.title,
    food_category: recipe.foodCategory,
    cooking_category: recipe.cookingCategory,
    images: {
      original_url: recipe.images.originalUrl,
      local_path: recipe.images.localPath,
    },
    tags: recipe.tags,
    tips: recipe.tips,
    ingredients: recipe.ingredients.map((i) => ({
      name: i.name,
      quantity: i.quantity,
    })),
    methods: recipe.methods.map((m) => ({
      describe: m.describe,
      image: {
        original_url: m.image.originalUrl,
        local_path: m.image.localPath,
      },
    })),
    nutrition: {
      calories: recipe.nutrition.calories,
      protein: recipe.nutrition.protein,
      carbohydrates: recipe.nutrition.carbohydrates,
      fat: recipe.nutrition.fat,
      sodium: recipe.nutrition.sodium,
    },
  };
}

export function recipeFromJson(json: Record<string, unknown>): Recipe {
  const imgs = json.images as Record<string, string>;
  const nutr = json.nutrition as Record<string, string>;
  return {
    id: json._id as string,
    title: json.title as string,
    foodCategory: json.food_category as string,
    cookingCategory: json.cooking_category as string,
    images: { originalUrl: imgs.original_url, localPath: imgs.local_path },
    tags: json.tags as string,
    tips: json.tips as string,
    ingredients: (json.ingredients as Record<string, string>[]).map((i) => ({
      name: i.name,
      quantity: i.quantity,
    })),
    methods: (json.methods as Record<string, unknown>[]).map((m) => {
      const img = m.image as Record<string, string>;
      return {
        describe: m.describe as string,
        image: { originalUrl: img.original_url, localPath: img.local_path },
      };
    }),
    nutrition: {
      calories: nutr.calories,
      protein: nutr.protein,
      carbohydrates: nutr.carbohydrates,
      fat: nutr.fat,
      sodium: nutr.sodium,
    },
  };
}
